XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs

```diff
@@ -132,23 +139,31 @@ private bool UploadChunk(string address, FileStream source, long offset, long th
 
             try
             {
-                var httpContent = new ByteArrayContent(buffer, 0, totalBytesRead);
-                httpContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
-                var response = httpClient.PostAsync(new Uri(address), httpContent, cancel).Result;
-                response.EnsureSuccessStatusCode();
-                var respString = response.Content.ReadAsStringAsync().Result;
-
-                // After sending every chunk upload request to server, response will contain a status indicating if it is complete.
-                var res = (Dictionary<string, object>)serializer.DeserializeObject(respString);
-                log.InfoFormat("The status of chunk upload: {0}", res.ContainsKey("status") ? res["status"] : "");
+                using (var httpContent = new ByteArrayContent(buffer, 0, totalBytesRead))
+                {
+                    httpContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
+                    using (var response = httpClient.PostAsync(new Uri(address), httpContent, cancel).Result)
+                    {
+                        if (response.IsSuccessStatusCode)
+                        {
+                            var respString = response.Content.ReadAsStringAsync().Result;
+
+                            // After sending every chunk upload request to server, response will contain a status indicating if it is complete.
+                            var res = (Dictionary<string, object>) serializer.DeserializeObject(respString);
+                            log.InfoFormat("The status of chunk upload: {0}", res.ContainsKey("status") ? res["status"] : "");
+                            return true;
+                        }
+                        log.ErrorFormat("Failed to upload the chunk. The exception was: {0} ({1})", response.StatusCode, response.ReasonPhrase);

```

Comment:
Strictly speaking this is not an exception, I'd rephrase the message to `Failed to upload the chunk. POST request returned {0} ({1})`.